proxy/src/outbound.rs

```diff
@@ -87,12 +88,20 @@ where
     fn recognize(&self, req: &Self::Request) -> Option<Reuse<Self::Key>> {
         let proto = bind::Protocol::detect(req);
 
-        let local = req.uri().authority_part().map(|authority| {
-            FullyQualifiedAuthority::normalize(
-                authority,
-                &self.default_namespace)
-
-        });
+        // The request URI and Host: header have not yet been normalized
+        // by `NormalizeUri`, as we need to know whether the request will
+        // be routed by Host/authority or by SO_ORIGINAL_DST, in order to
+        // determine whether the service is reusable.
+        let local = req.uri().authority_part()
+            .cloned()
+        // Therefore, we need to check the host header as well as the URI
+        // for a valid authority, before we fall back to SO_ORIGINAL_DST.
+            .or_else(|| h1::authority_from_host(req))
```

Comment:
If `h1::authority_from_host()` returns None then we should return None right here because `h1::authority_from_host()` does the `SO_ORIGNAL_DST` logic itself, so if it doesn't return an Authority then we have nothing to work with.

In particular, we don't need the `SO_ORIGINAL_DST` lower in this function any more, since ``h1::authority_from_host()` does that itself, right?

Reply:
As noted in #557, `h1::authority_from_host` does _not_ handle `SO_ORIGINAL_DST`.